File: backend/app/permissions.py
"""员工细粒度业务权限（管理员不受限）。"""

from __future__ import annotations

from typing import Any

PERM_ORDER_CREATE = "order_create"
PERM_ORDER_PROCESS = "order_process"
PERM_ORDER_OUTBOUND = "order_outbound"
PERM_ORDER_CONFIRM_SHIP = "order_confirm_ship"

PERMISSION_LABELS: dict[str, str] = {
    PERM_ORDER_CREATE: "新建订单",
    PERM_ORDER_PROCESS: "处理订单",
    PERM_ORDER_OUTBOUND: "出库订单",
    PERM_ORDER_CONFIRM_SHIP: "确认出库",
}

ALL_PERMISSION_CODES: tuple[str, ...] = tuple(PERMISSION_LABELS.keys())


def effective_permission_codes(user: Any) -> set[str]:
    if getattr(user, "role", None) == "admin":
        return set(ALL_PERMISSION_CODES)
    raw = getattr(user, "permission_codes", None)
    # NULL / 未配置：兼容旧员工，视为全部业务权限
    if raw is None:
        return set(ALL_PERMISSION_CODES)
    if not isinstance(raw, list):
        return set(ALL_PERMISSION_CODES)
    # 显式保存的空列表：无任何业务权限
    if len(raw) == 0:
        return set()
    return {str(x) for x in raw if x in PERMISSION_LABELS}


def has_permission(user: Any, code: str) -> bool:
    return code in effective_permission_codes(user)


def required_perm_for_item_patch(old: Any, data: dict) -> str:
    """单条明细 PATCH 所需业务权限码。"""
    if "production_status" not in data and "in_today_queue" not in data:
        return PERM_ORDER_PROCESS
    new_ps = data.get("production_status", old.production_status)
    if new_ps == "已发回":
        return PERM_ORDER_CONFIRM_SHIP
    old_ps = old.production_status
    if old_ps in ("待发回", "出库中"):
        return PERM_ORDER_OUTBOUND
    return PERM_ORDER_PROCESS


def required_perm_for_batch(items: list[Any], target_status: str) -> str:
    if target_status == "已发回":
        return PERM_ORDER_CONFIRM_SHIP
    if target_status in ("待发回", "出库中"):
        return PERM_ORDER_OUTBOUND
    for row in items:
        if row.production_status in ("待发回", "出库中"):
            return PERM_ORDER_OUTBOUND
    return PERM_ORDER_PROCESS

File: front/src/permissions.js
/** 与后端 app.permissions 一致 */
export const PERM = {
  ORDER_CREATE: 'order_create',
  ORDER_PROCESS: 'order_process',
  ORDER_OUTBOUND: 'order_outbound',
  ORDER_CONFIRM_SHIP: 'order_confirm_ship',
}

export const PERM_LABELS = {
  [PERM.ORDER_CREATE]: '新建订单',
  [PERM.ORDER_PROCESS]: '处理订单',
  [PERM.ORDER_OUTBOUND]: '出库订单',
  [PERM.ORDER_CONFIRM_SHIP]: '确认出库',
}

/** 管理员或未配置权限的员工：视为具备全部业务权限 */
export function can(user, code) {
  if (!user) return false
  if (user.role === 'admin') return true
  const raw = user.permission_codes
  if (!Array.isArray(raw) || raw.length === 0) return true
  return raw.includes(code)
}

export function canAnyOrderNav(user) {
  return (
    can(user, PERM.ORDER_CREATE) ||
    can(user, PERM.ORDER_PROCESS) ||
    can(user, PERM.ORDER_OUTBOUND) ||
    can(user, PERM.ORDER_CONFIRM_SHIP)
  )
}

export function canNavPending(user) {
  return can(user, PERM.ORDER_CREATE) || can(user, PERM.ORDER_PROCESS)
}

export function canNavProcessing(user) {
  return can(user, PERM.ORDER_PROCESS)
}

export function canNavReadyOutbound(user) {
  return can(user, PERM.ORDER_OUTBOUND) || can(user, PERM.ORDER_CONFIRM_SHIP)
}

export function canNavDone(user) {
  return can(user, PERM.ORDER_OUTBOUND) || can(user, PERM.ORDER_CONFIRM_SHIP)
}

/** 帐号管理里勾选顺序 */
export const PERM_OPTIONS = [
  [PERM.ORDER_CREATE, PERM_LABELS[PERM.ORDER_CREATE]],
  [PERM.ORDER_PROCESS, PERM_LABELS[PERM.ORDER_PROCESS]],
  [PERM.ORDER_OUTBOUND, PERM_LABELS[PERM.ORDER_OUTBOUND]],
  [PERM.ORDER_CONFIRM_SHIP, PERM_LABELS[PERM.ORDER_CONFIRM_SHIP]],
]
